### PR TITLE
Added API versioning support

### LINKED ISSUE #377: Add GitHub API versioning header
#### Feature Idea Summary
Add versioning to API request header.

#### Feature Idea Additional Details

GitHub has announced that they will start versioing their API. It's best to add the version to the API header. Using the 'X-GitHub-Api-Version' header and setting the value to the first version of their API '2022-11-28'.

https://github.blog/2022-11-28-to-infinity-and-beyond-enabling-the-future-of-githubs-rest-api-with-api-versioning/

#### Requested Assignment
If possible, I would like to implement this.
    

```
    $headers = @{
        'Accept' = $AcceptHeader
        'User-Agent' = 'PowerShellForGitHub'
        'X-GitHub-Api-Version'='2022-11-28'
    }
```

#### Operating System
OsName               : Microsoft Windows 10 Enterprise
OsOperatingSystemSKU : EnterpriseEdition
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 19041.1.amd64fre.vb_release.191206-1406
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

#### PowerShell Version
PSVersion                      7.2.6
PSEdition                      Core
GitCommitId                    7.2.6
OS                             Microsoft Windows 10.0.19044
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

#### Module Version
Running: 0.16.1
Installed: 0.16.1

### PR DESCRIPTION
**Pull request by variableresistor**
#### Description

GitHub is introducing versioned API's.  In order to get the expected behavior of an API, you need to invoke the API with its version in the request header.  This change updates the core functions (`Invoke-GHRestMethod` and `Invoke-GHRestMethodMultipleResult`) to accept an ApiVersion as part of their param sets so that as the module continues to add new functionality, methods can request the appropriate API Version for their individual calls.

#### Features Added

Fixes #377 

#### References

[To infinity and beyond: enabling the future of GitHub’s REST API with API versioning](https://github.blog/2022-11-28-to-infinity-and-beyond-enabling-the-future-of-githubs-rest-api-with-api-versioning/)
[API Versions](https://docs.github.com/en/rest/overview/api-versions?apiVersion=2022-11-28)

#### Checklist

- [x] You actually ran the code that you just wrote, especially if you did just "one last quick change".
- [x] Comment-based help added/updated, including examples.
- [x] [Static analysis](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#static-analysis) is reporting back clean.
- [x] New/changed code adheres to our [coding guidelines](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#coding-guidelines).
~~- [ ] [Formatters were created](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#formatters) for any new types being added.~~
- [x] New/changed code continues to [support the pipeline](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#pipeline-support).
- [x] Changes to the manifest file follow the [manifest guidance](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#module-manifest).
- [x] Unit tests were added/updated and are all passing. See [testing guidelines](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#testing).  This includes making sure that all pipeline input variations have been covered.
- [x] Relevant usage examples have been added/updated in [USAGE.md](https://github.com/microsoft/PowerShellForGitHub/blob/master/USAGE.md).
- [x] If desired, ensure your name is added to our [Contributors list](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#contributors)

### CODE REVIEW

**Comment by variableresistor** [comment]
> 

@microsoft-github-policy-service agree [company="City National Bank"]

**Comment by variableresistor** [comment]
> @variableresistor please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="City National Bank"

File: GitHubRepositories.ps1 [hunk 1]
```diff
@@ -68,6 +68,12 @@ filter New-GitHubRepository
     .PARAMETER NoWiki
         By default, this repository will have a Wiki.  Specify this to disable the Wiki.
 
+    .PARAMETER NoPages
```
Comment (HowardWolosky) [inline GitHubRepositories.ps1:71]: We should try to keep PR's as atomic as possible (which helps in cases like this where one part of the change is desired, but an unrelated change is contentious).  If you split this change out, we can likely get it merged in faster.

File: PowerShellForGitHub.psd1 [hunk 2]
```diff
@@ -7,7 +7,7 @@
     CompanyName = 'Microsoft Corporation'
     Copyright = 'Copyright (C) Microsoft Corporation.  All rights reserved.'
 
-    ModuleVersion = '0.16.1'
+    ModuleVersion = '0.17.0'
```
Comment (HowardWolosky) [inline PowerShellForGitHub.psd1:10]: Please do not change the module version yourself.  It only gets changed by project administrators at the time that an updated version is released to PowerShellGallery.

File: GitHubRepositories.ps1 [hunk 3]
```diff
@@ -1200,6 +1222,8 @@ filter Set-GitHubRepository
     if ($PSBoundParameters.ContainsKey('NoIssues')) { $hashBody['has_issues'] = (-not $NoIssues.ToBool()) }
     if ($PSBoundParameters.ContainsKey('NoProjects')) { $hashBody['has_projects'] = (-not $NoProjects.ToBool()) }
     if ($PSBoundParameters.ContainsKey('NoWiki')) { $hashBody['has_wiki'] = (-not $NoWiki.ToBool()) }
+    if ($PSBoundParameters.ContainsKey('NoPages')) { $hashBody['has_pages'] = (-not $NoPages.ToBool()) }
```
Comment (HowardWolosky) [inline GitHubRepositories.ps1:1225]: You may want to open a documentation bug on GitHub around this, as `has_pages` is not documented as a valid input, only as an expected output.

**Review by HowardWolosky — CHANGES_REQUESTED** [review]
Thanks for your contribution, @variableresistor!  I also appreciate you calling attention to the fact that GitHub is introducing API versioning...I completely missed that announcement.

I love the idea of making the module API version aware, but I currently disagree with the methodology that you're using to introduce it.  We're likely to adopt newer API versions on a per-API basis, one at a time.  That means we wouldn't want a module-wide mechanism for changing the API Version, but rather the ability to change it on every REST call.

File: GitHubCore.ps1 [hunk 4]
```diff
@@ -232,6 +232,7 @@ function Invoke-GHRestMethod
     $headers = @{
         'Accept' = $AcceptHeader
         'User-Agent' = 'PowerShellForGitHub'
+        'X-GitHub-Api-Version' = Get-GitHubConfiguration -Name "ApiVersion"
```
[the remark this answers is not in the document]
Comment (variableresistor) [reply GitHubCore.ps1:235]: Should I implement this as a simple string or in the datetime format, then ToString() into the yyyy-MM-dd date format?

File: GitHubConfiguration.ps1 [hunk 5]
```diff
@@ -71,6 +71,9 @@ function Set-GitHubConfiguration
         different hostname when using a GitHub Enterprise server. Do not include the HTTP/S prefix,
         and do not include 'api'. For example, use "github.contoso.com".
 
+    .PARAMETER ApiVersion
```
[the remark this answers is not in the document]
Comment (variableresistor) [reply GitHubConfiguration.ps1]: Removed

File: GitHubRepositories.ps1 (hunk 1 above)
Comment (variableresistor) [reply GitHubRepositories.ps1]: Reverted

File: PowerShellForGitHub.psd1 (hunk 2 above)
Comment (variableresistor) [reply PowerShellForGitHub.psd1]: Reverted

File: GitHubRepositories.ps1 (hunk 3 above)
Comment (variableresistor) [reply GitHubRepositories.ps1]: Reverted. Will open a documentation bug with the GitHub API people.

File: GitHubConfiguration.ps1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (variableresistor) [reply GitHubConfiguration.ps1]: @HowardWolosky is the routine in this repo to resolve these? Or would you like the PR author to resolve them once they've been corrected?

File: GitHubCore.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (HowardWolosky) [reply GitHubCore.ps1]: > Should I implement this as a simple string or in the datetime format, then ToString() into the yyyy-MM-dd date format?

Good question.  I think simple string makes the most sense here.  It remains the most flexible, and it also means that the calling code will be easier to read and write (e.g. `-ApiVersion '2022-11-28'` vs `-ApiVersion (Get-Date -Date '2022-11-28')`)

File: GitHubRepositories.ps1 (hunk 3 above)
Comment (HowardWolosky) [reply GitHubRepositories.ps1]: Awesome -- for tracking purposes, can you include a link to the bug report here after you've opened it?

File: GitHubCore.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (variableresistor) [reply GitHubCore.ps1]: Okay good, because that's how I implemented it. Then added the ValidatePattern validation attribute that checks for the correct format.

File: GitHubRepositories.ps1 (hunk 3 above)
Comment (variableresistor) [reply GitHubRepositories.ps1]: Checked into it, and their documentation is accurate. https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#update-a-repository.
That being said, I've opened an issue #378 to add Discussions support for Set-GitHubRepository (has_discussions property). I have an implementation for it once this PR is merged and closed to keep changes in my fork to a minimum.

**Review by HowardWolosky — CHANGES_REQUESTED** [review]
I thought I had published this review earlier, but it still shows as pending.

File: GitHubRepositories.ps1 (hunk 3 above)
Comment (variableresistor) [reply GitHubRepositories.ps1]: Right. That's their documentation is accurate and I am not 😅.

File: GitHubCore.ps1 [hunk 6]
```diff
@@ -101,6 +101,9 @@ function Invoke-GHRestMethod
         used as the exception bucket value in the event of an exception.  If neither is specified,
         no bucket value will be used.
 
+    .PARAMETER ApiVersion
```
[the remark this answers is not in the document]
Comment (variableresistor) [reply GitHubCore.ps1:104]: The same thought crossed my mind; that there's a better place for this parameter instead of just shoved at the end.

[the remark this answers is not in the document]
Comment (variableresistor) [reply GitHubCore.ps1]: Appended directly before AccessToken.

File: GitHubCore.ps1 [hunk 7]
```diff
@@ -663,7 +673,10 @@ function Invoke-GHRestMethodMultipleResult
 
         [string] $TelemetryExceptionBucket = $null,
 
-        [switch] $SinglePage
+        [switch] $SinglePage,
+
+        [ValidatePattern("\d\d\d\d-\d\d-\d\d")]
```
[the remark this answers is not in the document]
Comment (variableresistor) [reply GitHubCore.ps1]: Adjusted this regex pattern to mirror the other's

File: GitHubCore.ps1 [hunk 8]
```diff
@@ -151,7 +154,10 @@ function Invoke-GHRestMethod
 
         [hashtable] $TelemetryProperties = @{},
 
-        [string] $TelemetryExceptionBucket = $null
+        [string] $TelemetryExceptionBucket = $null,
+
+        [ValidatePattern("\d\d\d\d-\d\d-\d\d")]
+        [string]$ApiVersion = "2022-11-28"
```
[the remark this answers is not in the document]
Comment (variableresistor) [reply GitHubCore.ps1]: Mirrored regex with previous implementation

File: GitHubCore.ps1 [hunk 9]
```diff
@@ -632,6 +639,9 @@ function Invoke-GHRestMethodMultipleResult
         followed.
         WARNING: This might take a while depending on how many results there are.
 
+    .PARAMETER ApiVersion
```
[the remark this answers is not in the document]
Comment (variableresistor) [reply GitHubCore.ps1]: Relocated

File: GitHubCore.ps1 [hunk 10]
```diff
@@ -145,6 +148,9 @@ function Invoke-GHRestMethod
 
         [switch] $Save,
 
+        [ValidatePattern('^\d{4}-\d{2}-\d{2}$')]
+        [string]$ApiVersion = "2022-11-28",
```
Comment (HowardWolosky) [inline GitHubCore.ps1:152]: nit: spacing

```suggestion
        [string] $ApiVersion = "2022-11-28",
```

File: GitHubCore.ps1 [hunk 11]
```diff
@@ -660,6 +670,9 @@ function Invoke-GHRestMethodMultipleResult
 
         [hashtable] $AdditionalHeader = @{},
 
+        [ValidatePattern('^\d{4}-\d{2}-\d{2}$')]
+        [string]$ApiVersion = "2022-11-28",
```
Comment (HowardWolosky) [inline GitHubCore.ps1:674]: nit: spacing

```suggestion
        [string] $ApiVersion = "2022-11-28",
```

File: GitHubCore.ps1 [hunk 12]
```diff
@@ -703,6 +716,7 @@ function Invoke-GHRestMethodMultipleResult
                 'AccessToken' = $AccessToken
                 'TelemetryProperties' = $telemetryProperties
                 'TelemetryExceptionBucket' = $errorBucket
+                'ApiVersion' = $ApiVersion
```
Comment (HowardWolosky) [inline GitHubCore.ps1:719]: Let's keep this ordered the same way as the params:

```suggestion
                'ApiVersion' = $ApiVersion
                'AccessToken' = $AccessToken
                'TelemetryProperties' = $telemetryProperties
                'TelemetryExceptionBucket' = $errorBucket
```

File: GitHubCore.ps1 [hunk 13]
```diff
@@ -80,6 +80,9 @@ function Invoke-GHRestMethod
         If specified, this will save the result to a temporary file and return the FileInfo of that
         temporary file.
 
+    .PARAMETER ApiVersion
+        Version of the GitHub API to use. Format is: yyyy-MM-dd
```
Comment (HowardWolosky) [inline GitHubCore.ps1:84]:
```
        Indicates the version of this API that should be used. Format is: 'yyyy-MM-dd'.
```

File: GitHubCore.ps1 [hunk 14]
```diff
@@ -609,6 +616,9 @@ function Invoke-GHRestMethodMultipleResult
         Allows the caller to specify any number of additional headers that should be added to
         all of the requests made.
 
+    .PARAMETER ApiVersion
+        Version of the GitHub API to use. Format is: yyyy-MM-dd
```
Comment (HowardWolosky) [inline GitHubCore.ps1:620]:
```
        Indicates the version of this API that should be used. Format is: 'yyyy-MM-dd'.
```

**Review by HowardWolosky — CHANGES_REQUESTED** [review]
Super minor nits, and then should be good to merge.

File: GitHubCore.ps1 (hunk 12 above)
Comment (variableresistor) [reply GitHubCore.ps1:719]: Oh shoot. I thought I did that, committed that change.

File: GitHubCore.ps1 (hunk 13 above)
Comment (variableresistor) [reply GitHubCore.ps1]: Adjusted

File: GitHubCore.ps1 (hunk 14 above)
Comment (variableresistor) [reply GitHubCore.ps1]: Corrected parameter description

**Review by HowardWolosky — APPROVED** [review]
Looks good now.  Thanks for the contribution and for noticing this pending change in the first place!

Tests are sufficiently passing (pretty much all of the failing tests are the Projects tests which we know will fail (due to #380) -- they're about to be disabled).
